staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller_test.go

```diff
@@ -50,124 +63,268 @@ func TestProcessEncryptionConfig(t *testing.T) {
 			)
 
 			_, _, err := d.processEncryptionConfig(ctx)
-			if testCase.expectError && err == nil {
-				t.Fatalf("expected error but got none")
+
+			gotError := ""
+			if err != nil {
+				gotError = err.Error()
 			}
-			if !testCase.expectError && err != nil {
-				t.Fatalf("expected no error but got %v", err)
+
+			if gotError != testCase.expectError {
+				t.Fatalf("expected error %q but got %q", testCase.expectError, gotError)
 			}
 		})
 	}
 }
 
-func TestWatchEncryptionConfigFile(t *testing.T) {
-	testCases := []struct {
-		name          string
-		generateEvent func(filePath string, cancel context.CancelFunc)
-		expectError   bool
+// TestRun validates if Run is starting the controller and shutting it down when context is cancelled.
+func TestRun(t *testing.T) {
+	t.Run("to validate controller start and stop", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		d := NewDynamicEncryptionConfiguration(
+			"test",
+			"testdata/ec_config.yaml",
+			nil,
+			"",
+			"",
+		)
+
+		output := captureOutput(t, func() {
+			go func() {
+				d.Run(ctx)
+			}()
+
+			// allow the Run method to start
+			time.Sleep(time.Second)
+
+			cancel()
+
+			// allow the Run method to stop
+			time.Sleep(time.Second)
```

Comment:
@enj Is there a better way of doing this? I can't seem to find a way to reliably start and stop the controller for testing.

Reply:
same as https://github.com/kubernetes/kubernetes/pull/121310#discussion_r1367303187